File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/search/SearchHit.java
package fr.myprysm.vertx.elasticsearch.action.search;


import fr.myprysm.vertx.elasticsearch.action.support.DocumentField;
import fr.myprysm.vertx.elasticsearch.action.support.Explanation;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@DataObject(generateConverter = true)
public class SearchHit {

    private String id;
    private String type;
    private String index;
    private long version = -1;
    private JsonObject source;
    private Map<String, DocumentField> fields;
    private Map<String, List<String>> highlightFields;

    private List<String> matchedQueries;
    private Map<String, SearchHits> innerHits;
    private JsonArray sortValues;

    private NestedIdentity nestedIdentity;
    private Explanation explanation;

    private float score;
    private String clusterAlias;

    public SearchHit(SearchHit other) {
        id = other.id;
        type = other.type;
        index = other.index;
        version = other.version;
        source = other.source;
        fields = other.fields;
        highlightFields = other.highlightFields;
        matchedQueries = other.matchedQueries;
        innerHits = other.innerHits;
        sortValues = other.sortValues;
        nestedIdentity = other.nestedIdentity;
        explanation = other.explanation;
        score = other.score;
        clusterAlias = other.clusterAlias;
    }

    /**
     * Build a new <code>SearchHit</code> from a <code>JsonObject</code>.
     *
     * @param json the <code>JsonObject</code>
     */
    public SearchHit(JsonObject json) {
        SearchHitConverter.fromJson(json, this);
    }

    /**
     * Transforms the <code>SearchHit</code> into a <code>JsonObject</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = new JsonObject();
        SearchHitConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/search/SearchHit.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.search

import fr.myprysm.vertx.elasticsearch.action.search.SearchHit
import fr.myprysm.vertx.elasticsearch.action.search.NestedIdentity
import fr.myprysm.vertx.elasticsearch.action.search.SearchHits
import fr.myprysm.vertx.elasticsearch.action.support.DocumentField
import fr.myprysm.vertx.elasticsearch.action.support.Explanation

fun SearchHit(
        clusterAlias: String? = null,
        explanation: fr.myprysm.vertx.elasticsearch.action.support.Explanation? = null,
        fields: Map<String, fr.myprysm.vertx.elasticsearch.action.support.DocumentField>? = null,
        id: String? = null,
        index: String? = null,
        innerHits: Map<String, fr.myprysm.vertx.elasticsearch.action.search.SearchHits>? = null,
        matchedQueries: Iterable<String>? = null,
        nestedIdentity: fr.myprysm.vertx.elasticsearch.action.search.NestedIdentity? = null,
        score: Float? = null,
        sortValues: io.vertx.core.json.JsonArray? = null,
        source: io.vertx.core.json.JsonObject? = null,
        type: String? = null,
        version: Long? = null): SearchHit = fr.myprysm.vertx.elasticsearch.action.search.SearchHit().apply {

    if (clusterAlias != null) {
        this.setClusterAlias(clusterAlias)
    }
    if (explanation != null) {
        this.setExplanation(explanation)
    }
    if (fields != null) {
        this.setFields(fields)
    }
    if (id != null) {
        this.setId(id)
    }
    if (index != null) {
        this.setIndex(index)
    }
    if (innerHits != null) {
        this.setInnerHits(innerHits)
    }
    if (matchedQueries != null) {
        this.setMatchedQueries(matchedQueries.toList())
    }
    if (nestedIdentity != null) {
        this.setNestedIdentity(nestedIdentity)
    }
    if (score != null) {
        this.setScore(score)
    }
    if (sortValues != null) {
        this.setSortValues(sortValues)
    }
    if (source != null) {
        this.setSource(source)
    }
    if (type != null) {
        this.setType(type)
    }
    if (version != null) {
        this.setVersion(version)
    }
}

